### PR TITLE
Fix crash in Highway_overtake.wbt #3556 due to acos getting out of va…

### PR DESCRIPTION
**Pull request by llessieux**
Fix crash in Highway_overtake.wbt #3556 due to acos getting out of valid range value causing a NaN valid to be generated.

This in turn is assigned to a WbRotation angle which triggers an infinite notification loop leading to a stack overflow.
A equality check with a NaN value always being false, the WbSFRotation keeps on emitting changed signals.

Add WbMathsUtilities::clampedAcos and replace acos calls with that when the code in place was not already checking the bounds.
Add WbMathsUtilities::clampedAsin too, and update code using asin to prevent  indefinite values being generated.

WbRotation::isValid has been updated to check against NaN value stored in the angle part too.

Both clampedAcos and clampedAsin functions are inline for performance reasons.

### CODE REVIEW

File: src/webots/maths/WbMathsUtilities.hpp [hunk 1]
```diff
@@ -63,4 +67,25 @@ inline double WbMathsUtilities::normalizeAngle(double angle, double lastSpot = 0
     d -= 2.0 * M_PI;
   return d + lastSpot;
 }
+
+// Make sure that the value is within the valid range before calling acos
+// If not then the behavior is similar to a clamped value.
+inline double WbMathsUtilities::clampedAcos(double value) {
+  if (value >= 1.0)
+    return 0.0;
+  else if (value <= -1.0)
```
Comment (omichel) [inline src/webots/maths/WbMathsUtilities.hpp:76]: The `else` is not needed here.
```suggestion
  if (value <= -1.0)
```

File: src/webots/maths/WbMathsUtilities.hpp [hunk 2]
```diff
@@ -63,4 +67,25 @@ inline double WbMathsUtilities::normalizeAngle(double angle, double lastSpot = 0
...
+    return 2.0 * M_PI;
+  return acos(value);
+}
+
+// Make sure that the value is within the valid range before calling asin
+// If not then the behavior is similar to a clamped value.
+inline double WbMathsUtilities::clampedAsin(double value) {
+  if (value >= 1.0)
+    return M_PI / 2;
+  else if (value <= -1.0)
```
Comment (omichel) [inline src/webots/maths/WbMathsUtilities.hpp:86]:
```suggestion
  if (value <= -1.0)
```

**Comment by llessieux** [comment]
Something must be wrong with my Clang Format setup. It is complaining about an extra space but I can't figure out where...
Really strange. I made sure that all the files had been formatted before committing this time and yet...

File: src/webots/maths/WbRotation.hpp [hunk 3]
```diff
@@ -81,7 +81,8 @@ class WbRotation {
   }
 
   // invalid only if |axis| == 0.0
-  bool isValid() const { return !(mX == 0.0 && mY == 0.0 && mZ == 0.0); }
+  // the mAngle == mAngle checks the NaN case.
+  bool isValid() const { return !(mX == 0.0 && mY == 0.0 && mZ == 0.0) && (mAngle == mAngle); }
```
Comment (omichel) [inline src/webots/maths/WbRotation.hpp:85]: Interesting, I use to use `isnan(mAngle)` instead, but it is fine that way (and requires no dependency on math.h)

**Review by omichel — CHANGES_REQUESTED** [review]
That looks very good.
I will check the clang format problem.

File: src/webots/maths/WbRotation.hpp (hunk 3 above)
Comment (llessieux) [reply src/webots/maths/WbRotation.hpp:85]: yes I saw that after having committed the code that is was used in other places when I did the review.
Do you want me to replace it?

Comment (omichel) [reply src/webots/maths/WbRotation.hpp:85]: I believe it is indeed slightly more readable (will save us the comment).

**Comment by llessieux** [comment]
I should probably add something in the changelog I guess ;)

**Comment by omichel** [comment]
> I should probably add something in the changelog I guess ;)

Yes, please.

**Comment by ad-daniel** [comment]
I can confirm that I'm no longer having the crash on ubuntu 20 on this branch, thank you

File: docs/reference/changelog-r2021.md [hunk 4]
```diff
@@ -18,6 +18,7 @@ Released on XX, XXth, 2021.
     - Fixed mass calculation of Mesh nodes ([#3719](https://github.com/cyberbotics/webots/pull/3719)).
     - Fixed regression where the v3.3 (21 DoF) variant of the [Nao](../guide/nao.md) PROTO had no hands ([#3696](https://github.com/cyberbotics/webots/pull/3696)).
     - Fixed laser and infra-red distance sensors hitting fully transparent objects ([#3726](https://github.com/cyberbotics/webots/pull/3726)).
+    - Fixed a crash caused by acos function being called with out of valid range value and leading to a stack overflow. ([#3734](https://github.com/cyberbotics/webots/pull/3734))
```
Comment (omichel) [inline docs/reference/changelog-r2021.md:21]:
```suggestion
    - Fixed a crash caused by acos function being called with an out-of-range value and leading to a stack overflow ([#3734](https://github.com/cyberbotics/webots/pull/3734)).
```

File: src/webots/maths/WbRotation.hpp (hunk 3 above)
Comment (omichel) [inline src/webots/maths/WbRotation.hpp:85]:
```suggestion
  bool isValid() const { return !(mX == 0.0 && mY == 0.0 && mZ == 0.0) && !isnan(mAngle); }
```

**Review by omichel — CHANGES_REQUESTED** [review]

**Review by omichel — APPROVED** [review]
Thank you!

**Comment by llessieux** [comment]
Sorry it was dinner time here ;)
Merging now.

**Pull request merged**